### PR TITLE
[RFC] Using PMEMDevice as bluestore backend.

### PR DESCRIPTION
**Pull request by majianpeng**
Using libpmem(https://github.com/pmem/nvml/) to control pmemdevice as bluestore backend.

### CODE REVIEW

**Comment by majianpeng** [comment]
@liewegas, @yuyuyu101 . please review this. Thanks!

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -4104,46 +4104,54 @@ int BlueStore::_do_wal_op(bluestore_wal_op_t& wo, IOContext *ioc)
...
-	last.claim(first);   // same block we read above
-      } else {
-	dout(20) << __func__ << "  reading trailing partial block "
-		 << last_offset << "~" << block_size << dendl;
-	r = bdev->read(last_offset, block_size, &last, ioc, true);
-        assert(r == 0);
+#if defined(HAVE_PMEM)
+    if (bdev->get_type() == "pmem") {
+      //pmem don't need aligned w/ blocksize.
+      r = bdev->aio_write(offset, bl, ioc, true);
```
Comment (yuyuyu101) [inline src/os/bluestore/BlueStore.cc]: I'm not sure this change whether the potential constraint?

**Comment by yuyuyu101** [comment]
are you test with real device? nvdimm?

**Comment by majianpeng** [comment]
There is no real nvdimm device. I used ram to emulate a nvdimm device and created a dax-fs on this device to test. Now libpmem don't support raw nvdimm device(from the author of libpmem but i don't know why).But the author said it will be supported.

**Comment by zhouyuan** [comment]
FWIW, it's possible to do some quick tests with some simulated nvdimm device using the kernel parameter memmap

here's the detail instruction on this:
http://pmem.io/2016/02/22/pm-emulation.html

**Comment by majianpeng** [comment]
@liewegas . Hope your comments!

**Comment by liewegas** [comment]
@majianpeng can you rebase this on current master, and drop the last two commits?  (wal path is now simpler--just a simple overwrite, no read.)  You can also drop aio_zero. The rest of this looks pretty sane to me!

File: src/os/bluestore/BlockDevice.cc [hunk 2]
```diff
@@ -53,11 +60,42 @@ BlockDevice *BlockDevice::create(const string& path, aio_callback_t cb, void *cb
...
+      r = -errno;
+      derr << __func__ << " open " << path.c_str() << " error " << cpp_strerror(r) << dendl;
+    } else {
+      void *addr = ::mmap(0, 4096, PROT_WRITE, MAP_SHARED, fd, 0);
+      if (addr == MAP_FAILED) {
+	r = - errno;
+	derr << __func__ << " mmap " << path.c_str() << " error " << cpp_strerror(r) << dendl;
+      } else {
+	if (pmem_is_pmem(addr, 4096)) {
+	  type = "pmem";
```
Comment (liewegas) [inline src/os/bluestore/BlockDevice.cc]: dout(1) << __func__ << " auto-detected pmem device " << path << dendl;

File: src/os/bluestore/PMEMDevice.cc [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+  IOContext *ioc,
+  uint64_t offset,
+  uint64_t length)
+{
+  dout(20) << __func__  << " " << offset << "~" << length << dendl;
+  if (g_conf->bdev_debug_inflight_ios) {
+    Mutex::Locker l(debug_lock);
+    debug_inflight.erase(offset, length);
+  }
+}
```
Comment (liewegas) [inline src/os/bluestore/PMEMDevice.cc]: I think you can drop teh _aio_log_* functions entirely--nothing is aio here, ever.

File: src/os/bluestore/PMEMDevice.cc [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+  dout(40) << "data: ";
+  pbl->hexdump(*_dout);
+  *_dout << dendl;
+
+  _aio_log_finish(ioc, off, len);
+  --ioc->num_reading;
+  ioc->aio_wake();
+  return 0;
+}
```
Comment (liewegas) [inline src/os/bluestore/PMEMDevice.cc]: drop _aio_zero

File: src/os/bluestore/PMEMDevice.cc [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+int PMEMDevice::copy(uint64_t dest, uint64_t src, uint64_t len)
+{
+  dout(5) << __func__ << " from " << src  << " --> " << dest << " len=" << len << dendl;
+  assert(len > 0 );
+  assert(dest < size && src < size);
+  assert((src + len <= size) && (dest + len <= size));
+  pmem_memcpy_persist(addr + dest, addr + src, len);
+
+  return 0;
+}
```
Comment (liewegas) [inline src/os/bluestore/PMEMDevice.cc]: no need for copy

**Comment by dachary** [comment]
feel free to re-open once you have time to work on it

**Pull request closed without merging**